### ISSUE TITLE
train dataset structure?

### ISSUE THREAD

**Issue by sajjadaemmi** [PeikeLi/Self-Correction-Human-Parsing#14]
i want to train pascal model on own dataset. please explain me what is train dataset structure exactly?
annotations image pixels value must between 0 and 7 for each Person Part ??
what is train_id.txt?
thanks

**Comment by PeikeLi** [comment]
Exactly. Each pixel in the annotation image is the label_id. 
Each line in the trian_id.txt represents the img_name prefix of each image in the training set.
If you want to train on a customize dataset,  you need to also corresponding modifications for the dataset.py based on your own need (usually minor modification).

**Comment by sajjadaemmi** [comment]
can you explain what should i do for train step by step?
first run witch file? and then?
thanks

**Comment by VyBui** [comment]
@sajjadaemmi please follow the data structure of LIP (Look into people) dataset.

Simply put you need a train_images (original images) folder with train_id.txt accordingly mapping with the image's name. The same applies to validation_images.

However, for the train/Val segmentations there is a little bit trickier to get it. You have to have the labels (ground truth) of your definitions and then convert it to the class images before mapping its name to val_id.txt. 

The default LIP Val/train segmentations have 19 classes which mean each class represents a specific object.

Suppose your segmentation is the bellow image:

![0a0f64ffdb6aa45b0f445b217b05a6c6](https://user-images.githubusercontent.com/15938335/86097341-b3301400-bade-11ea-862f-9b963dae2833.png)

There are 11 classes in that color image, each one represents for an object in the picture (head, top, bottom .. etc. You can define your own objects/classes.)

Each class has a different color, for example, the hair has the value `(35, 125, 200)` in BGR. What you need to do is convert that specific color to the class.

Here is the snipped code that helps you convert the color to class:

`
COLOR_PICKER_THRESHOLD = 20
def convert_color_image_to_class_image(img: np.ndarray, LABEL_COLORS: np.array) -> np.ndarray:
 '''Args:
        img: Color image (np.adarray)

    Returns: Classes image
'''
    assert img is not None, "This is the empty image!"
    class_image = np.zeros((img.shape[0], img.shape[1], len(LABEL_COLORS)), np.uint8)

    class_image = np.zeros((img.shape[0], img.shape[1], len(LABEL_COLORS)), np.uint8)

    for i in range(1, len(LABEL_COLORS)):
        single_part = cv2.inRange(img, LABEL_COLORS[i] - COLOR_PICKER_THRESHOLD,
                                  LABEL_COLORS[i] + COLOR_PICKER_THRESHOLD)
        class_image[:, :, i] = single_part

    class_image = np.argmax(class_image, axis=2)
    class_image = np.uint8(class_image)

    return class_image
`

you define the variable 'LABEL_COLORS' like below:
`
CLASS_COLOR = {
 "hair": (0, 0, 0),
"other classes": (x, y, y)
}

LABEL_COLORS  = np.array([CLASS_COLOR['hair'],
CLASS_COLOR['other classes"']],
                                   dtype=np.float32)
`
Now the hair is the first class of the image which has the value of 0, then you feed both original image and segmentation image to the network and wait for the great and glorious work of @PeikeLi!

Thank @PeikeLi for make the training code available.

p/s: there is a minor issue in the training code that I have shared with @PeikeLi here: https://github.com/PeikeLi/Self-Correction-Human-Parsing/issues/17, go and fix it your self. 


regards,

**Comment by sajjadaemmi** [comment]
@PeikeLi and @VyBui 
thanks
